### PR TITLE
added search dropdown and routing to search pages

### PR DESCRIPTION
**Pull request by carsonwolber**
### What this pull request is doing
-  adds and styles the dropdown for search results when the topbar's search input is in focus mode
- allows for routing onto the 'search/url' page when a search result is clicked

### What will be done in later PRs and not included here

- currently the search isn't actually querying the database to show results matching the user's search. The PR following this will address this and make search fully functional. 

- PR after next will populate the search page. At minimum you'll be able to see a list of profiles matching your query. I would like to add an aggregation onto this page as well to give stats for the query object as a whole see: #23 

### This works towards (but doesn't complete) User Stories: 
- Users can search for other profiles 
- Users can search for other profiles through an array of filters/sorts on various metrics corresponding to profile attributes

## Test Plan

https://github.com/user-attachments/assets/bb56462c-c8f8-4506-bb36-a8a23e5a2326

I manually tested different edge case inputs for `searchRouteFormatter` as well but it'd be a good enhancement to add an explicit unit testing suite to give further confidence it works (maybe this should be an issue too)

### CODE REVIEW

File: frontend/src/HomePage/Top/Search.jsx [hunk 1]
```diff
@@ -1,18 +1,64 @@
...
-            <div className='icon'>
-                <svg xmlns="http://www.w3.org/2000/svg" width="20" height="16" fill="#F7F9FB" className="bi bi-search" viewBox="0 0 16 16">
-                <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001q.044.06.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1 1 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0"/>
-                </svg>
-            </div>
+    const [query, setQuery] = useState('');
+    const navigate = useNavigate();
+
+    // when a user first focuses on the search bar, they should see some default queries
+    const defaultSeach =  [{label: 'Menlo Park, CA', category: 'city' },
```
Comment (William-La) [inline frontend/src/HomePage/Top/Search.jsx]:
```suggestion
    const defaultSearch =  [{label: 'Menlo Park, CA', category: 'city' },
```

typo

File: frontend/HelperFuncs/utils.js [hunk 2]
```diff
@@ -219,4 +219,31 @@ export function isValidPassword(password){
...
+
+/*
+helper function to convert the dictionary form of a search object selected by a user into a 
+readable URL endpoint for navigaion. Each category adheres to the following unique form factors: 
+city: 'Menlo Park, CA' -> 'menloparkCA'
+salary: '$87,076 - $170,050' -> '87076-170050'
+job: 'Software Enigneer' -> 'SoftwareEngineer'
+requires [search] to be a dictionary with fields label and category 
+*/
+export function searchRouteFormatter(search) {
```
Comment (William-La) [inline frontend/HelperFuncs/utils.js:233]: how will we handle unexpected characters?

**Comment by William-La** [comment]
from your video it looks like the search dropdown is animated to come in from the side rather than top down-- seems a bit unnatural.

**Review by William-La — APPROVED** [review]
Overall looks good, added a few nit comments

File: frontend/src/HomePage/Top/Search.jsx (hunk 1 above)
Comment (carsonwolber) [reply frontend/src/HomePage/Top/Search.jsx]: fixed  in 9983964

File: frontend/HelperFuncs/utils.js (hunk 2 above)
Comment (carsonwolber) [reply frontend/HelperFuncs/utils.js:233]: This is a good point, although job title characters will in theory never be truly unexpected since the number of job titles is finite thanks to the use of the API (controlling for these kinds of scenarios is exactly why I though implementing the API was worthwhile). For example, one of the more edge-case looking job titles in the database 'An/Ssn-2 (V) 4 Operator' gets routed to the page `http://localhost:5173/search/An/Ssn-2(V)4Operator` which is an issue because the additional `/` in the job title might cause some trouble. I'll make a commit to adjust the function to use `encodeURIComponent` which should cover this and allow for better spacing than camel case.

**Comment by carsonwolber** [comment]
> from your video it looks like the search dropdown is animated to come in from the side rather than top down-- seems a bit unnatural.

This was actually directly inspired by [level.fyis](https://www.levels.fyi/?compare=Amazon,Cisco,Optum&track=Software%20Engineer) design since their search behaves similarly. The thing I'm more concerned with design wise is that it's a little slow because I had issues getting the click onto an li element to register with a faster TimeOut for `onBlur`. I did find it can be set down to 125 from 150 and still work and I think that cleans it up a bit so I'll commit that change as well.

**Comment by carsonwolber** [comment]
> > from your video it looks like the search dropdown is animated to come in from the side rather than top down-- seems a bit unnatural.
> 
> This was actually directly inspired by [level.fyis](https://www.levels.fyi/?compare=Amazon,Cisco,Optum&track=Software%20Engineer) design since their search behaves similarly. The thing I'm more concerned with design wise is that it's a little slow because I had issues getting the click onto an li element to register with a faster TimeOut for `onBlur`. I did find it can be set down to 125 from 150 and still work and I think that cleans it up a bit so I'll commit that change as well.

fixed in 3896227

File: frontend/HelperFuncs/utils.js (hunk 2 above)
Comment (carsonwolber) [reply frontend/HelperFuncs/utils.js:233]: Screenshot 2024-07-12 at 10 42 03 AM


this should now be fixed. in 96fee21
